Validation/CTPPS/plugins/CTPPSDirectProtonSimulation.cc

```diff
@@ -0,0 +1,500 @@
+/****************************************************************************
+ * Authors:
+ *   Jan Kašpar
+ *   Laurent Forthomme
+ ****************************************************************************/
+
+
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+#include "FWCore/Framework/interface/stream/EDProducer.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/Framework/interface/ESHandle.h"
+#include "FWCore/Framework/interface/ESWatcher.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+
+#include "SimDataFormats/GeneratorProducts/interface/HepMCProduct.h"
+
+#include "DataFormats/CTPPSDetId/interface/CTPPSDetId.h"
+#include "DataFormats/CTPPSDetId/interface/CTPPSPixelDetId.h"
+
+#include "DataFormats/Common/interface/DetSetVector.h"
+#include "DataFormats/CTPPSReco/interface/TotemRPRecHit.h"
+#include "DataFormats/CTPPSReco/interface/CTPPSDiamondRecHit.h"
+#include "DataFormats/CTPPSReco/interface/CTPPSPixelRecHit.h"
+#include "DataFormats/CTPPSReco/interface/CTPPSLocalTrackLite.h"
+
+#include "CondFormats/RunInfo/interface/LHCInfo.h"
+#include "CondFormats/DataRecord/interface/LHCInfoRcd.h"
+
+#include "CondFormats/DataRecord/interface/CTPPSOpticsRcd.h"
+#include "CondFormats/CTPPSReadoutObjects/interface/LHCOpticalFunctionsCollection.h"
+
+#include "CondFormats/CTPPSReadoutObjects/interface/CTPPSBeamParameters.h"
+#include "CondFormats/DataRecord/interface/CTPPSBeamParametersRcd.h"
+
+#include "Geometry/VeryForwardGeometryBuilder/interface/CTPPSGeometry.h"
+#include "Geometry/Records/interface/VeryForwardMisalignedGeometryRecord.h"
+#include "Geometry/VeryForwardRPTopology/interface/RPTopology.h"
+
+#include <unordered_map>
+
+#include "TMath.h"
+#include "TMatrixD.h"
+#include "TVectorD.h"
+
+//----------------------------------------------------------------------------------------------------
+
+class CTPPSDirectProtonSimulation : public edm::stream::EDProducer<>
```

Comment:
Unfortunately "official" PPS fast simulation uses and different geometry, proton transport, etc. In order to avoid effects of these differences, the validation suite for proton reconstruction contains this simple module.